client/src/components/ElementActions/ArchiveAction.js

```diff
@@ -0,0 +1,46 @@
+/* global confirm */
+import React from 'react';
+import { compose } from 'redux';
+import AbstractAction from 'components/ElementActions/AbstractAction';
+import archiveBlockMutation from 'state/editor/archiveBlockMutation';
+import i18n from 'i18n';
+
+/**
+ * Adds the elemental menu action to archive a block of any state
+ */
+const ArchiveAction = (WrappedComponent) => (props) => (
+  <WrappedComponent {...props}>
+    {props.children}
+
+    <AbstractAction
+      title={i18n._t('ArchiveAction.ARCHIVE', 'Archive')}
+      extraClass="element-editor__actions-archive"
+      onClick={(event) => {
+        event.stopPropagation();
+
+        const { id, isPublished, actions: { handleArchiveBlock } } = props;
+
+        let archiveMessage = i18n._t(
+          'ArchiveAction.CONFIRM_DELETE',
+          'Are you sure you want to send this block to the archive?'
+        );
+
+        if (isPublished) {
+          archiveMessage = i18n._t(
+            'ArchiveAction.CONFIRM_DELETE_AND_UNPUBLISH',
+            'Warning: This block will be unpublished before being sent to the archive. Are you sure you want to proceed?'
+          );
+        }
+
+        // eslint-disable-next-line no-alert
+        if (handleArchiveBlock && confirm(archiveMessage)) {
+          handleArchiveBlock(id);
+        }
+      }}
+    />
+  </WrappedComponent>
+);
```

Comment:
It took us a bit of time to get our head around this. Partially because of `WrappedComponent` vs `WrappingComponent`. Dylan has suggested perhaps a nicer way to handle this, which gives the wrapped "`MenuComponent`" more flexibility:

```js
const HOC = (MenuComponent) => (props) => {
	const newProps = {
		...props,
		actions: [
			...props.actions,
			<AbstractAction {...} /> OR { title: 'thing', handleClick: () => {...} } // <- Shaped object type
		]
	};

	return <MenuComponent {...newProps} />
}
```

Edit tabs could be added to the `MenuComponent` actions too, and perhaps something for separators too - reactstrap has a `<Separator>` component for separators, and if we go the "shaped object type" approach we could have nested arrays?

This gives MenuComponent more flexibility to choose how actions render. We'd like to hear your thoughts.

Reply:
Yeah I'd like to use the AbstractAction as much as possible for this menu - I've already added to the method that renders the CMS tab actions

Reply:
I've pushed a new commit - let me know if that satisfies your suggestion :)